src/re/multi_grid/indexing.py

```diff
@@ -0,0 +1,969 @@
+# SPDX-License-Identifier: GPL-2.0+ OR BSD-2-Clause
+# Authors: Gordian Edenhofer, Philipp Frank
+#!/usr/bin/env python3
+
+import operator
+from dataclasses import dataclass
+from functools import partial, reduce
+from typing import Callable, Iterable, Optional
+
+import jax.numpy as jnp
+import numpy as np
+import numpy.typing as npt
+from jax import ShapeDtypeStruct, eval_shape, vmap
+from jax.experimental import checkify
+from jax.lax import select
+
+from ..tree_math.pytree_string import PyTreeString
+from .jhealpix_reference import get_all_neighbours_valid, pix2vec, vec2pix
+
+
+@dataclass()
+class GridAtLevel:
+    shape: npt.NDArray[np.int_]
+    splits: Optional[npt.NDArray[np.int_]]
+    parent_splits: Optional[npt.NDArray[np.int_]]
+
+    def __init__(self, shape, splits=None, parent_splits=None):
+        self.shape = np.atleast_1d(shape)
+        if splits is not None:
+            splits = np.atleast_1d(splits)
+        self.splits = splits
+        if parent_splits is not None:
+            parent_splits = np.atleast_1d(parent_splits)
+        self.parent_splits = parent_splits
+
+    def _parse_index(self, index):
+        index = jnp.asarray(index)
+        if index.shape[0] != self.shape.size:
+            l = index.shape[0]
+            ve = f"index {index} is of invalid length {l} for shape {self.shape}"
+            raise IndexError(ve)
+        # Follow jax-array style out of bounds handling
+        shp_bc = self.shape[(slice(None),) + (np.newaxis,) * (index.ndim - 1)]
+        index = select(jnp.abs(index) < shp_bc, index, jnp.sign(index) * (shp_bc - 1))
+        return index % shp_bc
+
+    @property
+    def size(self):
+        return reduce(operator.mul, self.shape, 1)
+
+    @property
+    def ndim(self):
+        return len(self.shape)
+
+    def refined_indices(self):
+        if self.splits is None:
+            raise IndexError("This level has no children")
+        # TODO non-dense grid
+        return np.mgrid[tuple(slice(0, sh) for sh in self.shape)]
+
+    def resort(self, batched_ar):
```

Comment:
Does not need to be part of the grid structure? It enforces a specific memory layout that is not needed anywhere in the grid, I think.